localstack/services/generic_proxy.py

```diff
@@ -592,136 +594,88 @@ def get_flask_ssl_context(cls, serial_number=None):
-        self.base_url = urlparse(base_url)
-        self.forward_url = urlparse(forward_url)
-
-    def forward_request(self, method, path, data, headers):
-        host = headers.get("Host", "")
-
-        if not self.matches(host, path):
-            return True
-
-        # build forward url
-        forward_url = self.build_forward_url(host, path)
-
-        # update headers
-        headers.pop("Host", None)
-        headers["Host"] = forward_url.netloc
-
-        # TODO: set proxy headers like x-forwarded-for?
-
-        return self.do_forward(method, forward_url.geturl(), headers, data)
-
-    def do_forward(self, method, url, headers, data):
-        return requests.request(method, url, data=data, headers=headers, stream=True, verify=False)
+    def __init__(self, base_url):
+        self._url = urlparse(base_url)
+        super().__init__(self._url.port, f"{self._url.hostname}")
+        self.rules = []
 
-    def matches(self, host, path):
-        # TODO: refine matching default ports (80, 443 if scheme is https). Example:
-        #  http://localhost:80 matches
-        #  http://localhost) check host rule. Can lead to problems with 443-4566 edge proxy
-        #  forwarding if not enabled
-        if self.base_url.netloc:
-            stripped_netloc, _, port = self.base_url.netloc.rpartition(":")
-            if host != self.base_url.netloc and (
-                host != stripped_netloc or port not in ["80", "443"]
-            ):
-                return False
-
-        # check path components
-        if self.base_url.path == "/":
-            if path.startswith("/"):
-                return True
-
-        path_parts = path.split("/")
-        base_path_parts = self.base_url.path.split("/")
-
-        if len(base_path_parts) > len(path_parts):
-            return False
-
-        for i, component in enumerate(base_path_parts):
-            if component != path_parts[i]:
-                return False
-
-        return True
-
-    def build_forward_url(self, host, path):
-        # build forward url
-        if self.forward_url.hostname:
-            forward_host = self.forward_url.scheme + "://" + self.forward_url.netloc
-        else:
-            forward_host = host
-        forward_path_root = self.forward_url.path
-        forward_path = path[len(self.base_url.path) :]  # strip base path
-
-        # avoid double slashes
-        if forward_path and not forward_path_root.endswith("/"):
-            if not forward_path.startswith("/"):
-                forward_path = "/" + forward_path
-
-        forward_url = forward_host + forward_path_root + forward_path
+    def register(self, forward_url):
+        """
+        Registers two routes at the edge router to implement either a path or a domain based routing.
+        We need two rules to match 1) the root path + no additional path 2) the root path + any other path.
+        This function is NOT called automatically, it is usually called in the leaf class' start up routine.
```

Comment:
This comment raises a flag for me (since you document that something that one would expect to happen automatically does not happen automatically).
Isn't there a chance to do this automatically? If not, what's the point in having this in the class hierarchy?

Reply:
It doesn't happen automatically because the Proxy usually doesn't know at creation time where the cluster it manages resides, only after actually starting the cluster this is assigned. The alternative I see right now is reserving the port preemptively (which is justifiable I guess, since without starting the cluster there is little value in using the service). The hostname is know on startup.